ext/dom/document.c

```diff
@@ -1039,18 +1039,68 @@ PHP_METHOD(DOMDocument, getElementById)
 }
 /* }}} end dom_document_get_element_by_id */
 
+static void php_dom_transfer_document_ref(xmlNodePtr node, dom_object *dom_object_document, xmlDocPtr document)
+{
+	if (node->children) {
+		php_dom_transfer_document_ref(node->children, dom_object_document, document);
+	}
+	while (node) {
+		php_libxml_node_ptr *iteration_object_ptr = node->_private;
+		if (iteration_object_ptr) {
+			php_libxml_node_object *iteration_object = iteration_object_ptr->_private;
+			ZEND_ASSERT(iteration_object != NULL);
+			dom_object_document->document->refcount++;
+			 php_libxml_decrement_doc_ref(iteration_object);
+			iteration_object->document = dom_object_document->document;
+			// php_libxml_increment_doc_ref(iteration_object, document);
+		}
+		node = node->next;
+	}
+}
+
 /* {{{ URL: http://www.w3.org/TR/2003/WD-DOM-Level-3-Core-20030226/DOM3-Core.html#core-Document3-adoptNode
 Since: DOM Level 3
+Modern spec URL: https://dom.spec.whatwg.org/#dom-document-adoptnode
 */
 PHP_METHOD(DOMDocument, adoptNode)
 {
-	zval *nodep = NULL;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "O", &nodep, dom_node_class_entry) == FAILURE) {
+	zval *node_zval;
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "O", &node_zval, dom_node_class_entry) == FAILURE) {
 		RETURN_THROWS();
 	}
 
-	DOM_NOT_IMPLEMENTED();
+	xmlNodePtr nodep;
+	dom_object *dom_object_nodep;
+	DOM_GET_OBJ(nodep, node_zval, xmlNodePtr, dom_object_nodep);
+
+	if (UNEXPECTED(nodep->type == XML_DOCUMENT_NODE
+		|| nodep->type == XML_DOCUMENT_TYPE_NODE
+		|| nodep->type == XML_HTML_DOCUMENT_NODE
+		|| nodep->type == XML_ENTITY_NODE
+		|| nodep->type == XML_NOTATION_NODE)) {
+		php_dom_throw_error(NOT_SUPPORTED_ERR, true);
+		return;
```

Comment:
As this always seems to throw
```suggestion
		RETURN_THROWS();
```